### ISSUE TITLE
TOML language support (via taplo) effectively disabled by default.

### ISSUE THREAD

**Issue by gdtroszak** [helix-editor/helix#3897]
### Summary

I installed taplo for TOML language support according to the [instructions on the wiki](https://github.com/helix-editor/helix/wiki/How-to-install-the-default-language-servers#toml). When opening a TOML file, there is a single problem reported at the first character in the file - `this document has been excluded`. This is coming from `taplo`, and appropriately crops up when a file is [excluded](https://taplo.tamasfe.dev/configuration/file.html#exclude) from taplo evaluation (or doesn't match an [include](https://taplo.tamasfe.dev/configuration/file.html#include) configuration).

I was able to fix this by adding this block to my `languages.toml`:

```toml
[[language]]
name = 'toml'
config = {}
```

Is this the expected default behavior? If so, perhaps this should be noted in the [documentation for installing the toml language server](https://github.com/helix-editor/helix/wiki/How-to-install-the-default-language-servers#toml)? If not, perhaps the [default languages.toml](https://github.com/helix-editor/helix/blob/master/languages.toml) should be updated to include the `config` block? Either way, I'm happy to open a PR to try to resolve.

### Reproduction Steps

 

I tried this:

1. `cargo install taplo-cli --locked --features lsp`
2. `hx some_file.toml`

I expected this to happen:

TOML language support.

Instead, this happened:

Only shown a single issue at the first character in the TOML file `this document has been excluded`.

### Helix log

<details><summary>~/.cache/helix/helix.log</summary>

```
2022-09-18T19:37:33.981 helix_lsp::transport [ERROR] err <- " INFO taplo: registered request handler method=\"initialize\"\n"
2022-09-18T19:37:33.981 helix_lsp::transport [ERROR] err <- " INFO taplo: registered request handler method=\"textDocument/foldingRange\"\n"
2022-09-18T19:37:33.981 helix_lsp::transport [ERROR] err <- " INFO taplo: registered request handler method=\"textDocument/documentSymbol\"\n"
2022-09-18T19:37:33.981 helix_lsp::transport [ERROR] err <- " INFO taplo: registered request handler method=\"textDocument/completion\"\n"
2022-09-18T19:37:33.981 helix_lsp::transport [ERROR] err <- " INFO taplo: registered request handler method=\"textDocument/formatting\"\n"
2022-09-18T19:37:33.981 helix_lsp::transport [ERROR] err <- " INFO taplo: registered request handler method=\"textDocument/hover\"\n"
2022-09-18T19:37:33.981 helix_lsp::transport [ERROR] err <- " INFO taplo: registered request handler method=\"textDocument/documentLink\"\n"
2022-09-18T19:37:33.981 helix_lsp::transport [ERROR] err <- " INFO taplo: registered request handler method=\"textDocument/semanticTokens/full\"\n"
2022-09-18T19:37:33.981 helix_lsp::transport [ERROR] err <- " INFO taplo: registered request handler method=\"textDocument/prepareRename\"\n"
2022-09-18T19:37:33.981 helix_lsp::transport [ERROR] err <- " INFO taplo: registered request handler method=\"textDocument/rename\"\n"
2022-09-18T19:37:33.981 helix_lsp::transport [ERROR] err <- " INFO taplo: registered notification handler method=\"initialized\"\n"
2022-09-18T19:37:33.981 helix_lsp::transport [ERROR] err <- " INFO taplo: registered notification handler method=\"textDocument/didOpen\"\n"
2022-09-18T19:37:33.981 helix_lsp::transport [ERROR] err <- " INFO taplo: registered notification handler method=\"textDocument/didChange\"\n"
2022-09-18T19:37:33.981 helix_lsp::transport [ERROR] err <- " INFO taplo: registered notification handler method=\"textDocument/didSave\"\n"
2022-09-18T19:37:33.981 helix_lsp::transport [ERROR] err <- " INFO taplo: registered notification handler method=\"textDocument/didClose\"\n"
2022-09-18T19:37:33.981 helix_lsp::transport [ERROR] err <- " INFO taplo: registered notification handler method=\"workspace/didChangeConfiguration\"\n"
2022-09-18T19:37:33.981 helix_lsp::transport [ERROR] err <- " INFO taplo: registered notification handler method=\"workspace/didChangeWorkspaceFolders\"\n"
2022-09-18T19:37:33.981 helix_lsp::transport [ERROR] err <- " INFO taplo: registered request handler method=\"taplo/convertToJson\"\n"
2022-09-18T19:37:33.981 helix_lsp::transport [ERROR] err <- " INFO taplo: registered request handler method=\"taplo/convertToToml\"\n"
2022-09-18T19:37:33.981 helix_lsp::transport [ERROR] err <- " INFO taplo: registered request handler method=\"taplo/listSchemas\"\n"
2022-09-18T19:37:33.981 helix_lsp::transport [ERROR] err <- " INFO taplo: registered request handler method=\"taplo/associatedSchema\"\n"
2022-09-18T19:37:33.981 helix_lsp::transport [ERROR] err <- " INFO taplo: registered notification handler method=\"taplo/associateSchema\"\n"
2022-09-18T19:37:33.981 helix_lsp::transport [ERROR] err <- " INFO taplo: LSP server listening transport=\"stdio\"\n"
2022-09-18T19:37:33.982 helix_lsp::transport [ERROR] err <- " WARN taplo:document_open: using detached workspace document_url=file://~/.config/helix/languages.toml\n"
```

</details>

### Platform

macOS

### Terminal Emulator

Alacritty 0.10.0

### Helix Version

helix 22.08.1 (66276ce6)

**Comment by David-Else** [comment]
It is working OK for me on Linux using the binary: https://github.com/tamasfe/taplo/releases/download/release-taplo-cli-0.7.0/taplo-full-x86_64-unknown-linux-gnu.tar.gz
```
taplo --version
taplo 0.7.0
```
I have not touched the default `languages.toml`

**Comment by dead10ck** [comment]
Closing in favor of #3901. Thanks for your report!

**Comment by Zykino** [comment]
> It's not a problem, it only returns that message because taplo does not have a schema for `languages.toml` and `config.toml` which is helix specific.
> 
> [Taplo Schemas](https://taplo.tamasfe.dev/schema_index.json) [Json SchemaStore](https://www.schemastore.org/json/)

What? Taplo need to have the schema in its own config file (your first link)? Even if the schema I want is on the 2nd link?
Can helix parse the TOML file and "configure" taplo with it?

I mean, following [starship’s config example](https://starship.rs/config/#prompt) I though that adding the schema at the begining of the file was enough:
```toml
# Get editor completions based on the config schema
"$schema" = 'https://starship.rs/config-schema.json'

# Rest of their sample config …
```

**Comment by gdtroszak** [comment]
FWIW, I still had the issue opening up TOML files that weren't helix's `config.toml` or `languages.toml` until I added the blanket `config = {}` alluded to in my original post. It seems odd that adding taplo schema's for those specific file types would fix the general issue of just totally ignoring all TOML files. That said, I know nothing about the inner workings of either helix or taplo, so could totally be missing something.

**Comment by Zykino** [comment]
Well on my helix version `helix 22.05 (bdd1192a)` with latest `taplo 0.7.0` installed with `cargo`, adding the `config = {}` in the toml’s language section still tells me that `~/.config/starship.toml` is excluded.

To me it is not a big annoyance as I learned about toml’s lsp just know. But I wanted to try and that was not good :(.
I tried to opened a `Cargo.toml` and it worked :tada: so at least for the file I will use the most I will be able to see what all this have to offer for a data format file :).

**Comment by dbarnett** [comment]
@GitMurf Thank you! I finally got it working after adding this to my languages.toml:

```toml
[language-server.taplo.config]
# Workaround for nonsense of being nonfunctional outside git directories
# (https://github.com/helix-editor/helix/issues/3897#issuecomment-2361674437)
root_dir = [".git", "*.toml"]
```

I followed up in tamasfe/taplo#439 since I couldn't find any better canonical issue. AFAICT the primary issue is just that taplo LSP's "detached workspace" mode when it can't find a workspace root is pretty badly broken.

**Comment by GitMurf** [comment]
@dbarnett curious are you on windows? I am on windows and I am wondering if it's a dreaded file path format issue like 95% of my problems are? 🤔

**Comment by dbarnett** [comment]
Nope, Linux. This seems to be a proudly cross-platform bug in taplo. 😆

**Comment by Huliiiiii** [comment]
Although adding all of the above to my language.toml, taplo still shows "this document has been excluded". I can format them via taplo cli, but not in helix. Does anyone know how to fix this on Windows?

**Comment by dbarnett** [comment]
The important thing is to make taplo find a "workspace root" in the parent directories of whatever file you are editing. The above config isn't ideal but I don't know how it could fail to work around the problem, because it makes the immediate parent dir of any .toml file you edit a "workspace root". Are you sure the language.toml file you're editing is actually being picked up by helix?

**Comment by Huliiiiii** [comment]
Yes. I think taplo did not successfully find the workspace root on Windows. As for the other toml files under git, everything is fine.

**Comment by danielwe** [comment]
> ```toml
> [language-server.taplo.config]
> # Workaround for nonsense of being nonfunctional outside git directories
> # (https://github.com/helix-editor/helix/issues/3897#issuecomment-2361674437)
> root_dir = [".git", "*.toml"]
> ```

I went down the rabbit hole on this one and realized that the `root_dir` key isn't doing anything. The available config keys for taplo are `cache_path` and `configuration_section`, as you can see here: https://github.com/tamasfe/taplo/blob/1bb2c1881995a819d2bda7681d145709c783be1b/crates/taplo-lsp/src/config.rs#L12-L18. Other keys are simply ignored. It seems like the difference here is simply whether a `config` table exists at all---if it does, taplo's detached workspace mode works, otherwise you get the "document has been excluded" message and the LSP shuts down. Why, I have no idea, I suppose the fix would have to be in taplo's initialization code.

In short, all you need is the following:
```toml
[language-server.taplo]
config = {}
```

**Comment by Zykino** [comment]
Just found out that nothing happend anymore for me. And the log showed:

> 2025-05-18T23:26:12.710 helix_lsp::transport [ERROR] taplo err: <- StreamClosed
> 2025-05-18T23:26:12.710 helix_lsp [ERROR] failed to initialize language server: server closed the stream

Apparently `taplo-cli` does not come with its lsp anymore so had to re-install. In my case from cargo: `cargo install taplo-cli --locked --features lsp` and your config made it work (at least on "languages.toml" which did not work before).

Thanks :)

**Comment by Susensio** [comment]
> In short, all you need is the following:
> 
> [language-server.taplo]
> config = {}

shouldn't this be in the defaults?
@dead10ck this is not a duplicated of the issue you linked, would you mind to reopen?
